### PR TITLE
Typescript rewrite

### PR DESCRIPTION
**Pull request by tovade**
This PR includes

- DynoxHost rewritten in typescript
- typings removed because typescript makes them when building the files
- when publishing to npm dont forget to do `tsc`
more questions? let me know!

### CODE REVIEW

**Comment by devnote-dev** [comment]
Since this is being rewritten, a few things can be fixed in this PR too :P

File: src/DynoxHost.ts [hunk 1]
```diff
@@ -1,110 +1,110 @@
...
+        if (!api) throw new Error('[DH.js] Missing User API Key');
+        if (typeof api !== 'string') throw new Error('[DH.js] User API Key must be a String');
+        this.apikey = api
+
+    };
+
+    async getUsage(id: string) {
+        /**
+         * Get server usage/stats.
+         * @param {String} id Server ID
```
Comment (devnote-dev) [inline src/DynoxHost.ts]: JSDocs are meant to be before declaration :P

File: src/DynoxHost.ts [hunk 2]
```diff
@@ -1,110 +1,110 @@
...
+        if (res.status === 404) throw new Error(`[DH.js] Server (${id}) not found`);
+        if (res.status !== 200) throw new Error(`[DH.js] An unexpected error has occurred`);
+        let json = res.json();
+        return json;
+    };
+
+    async getDetails(id: string) {
+        /**
+         * Get server details.
+         * @param {String} id Server ID
```
Comment (devnote-dev) [inline src/DynoxHost.ts]: same here

**Comment by LogischJo** [comment]
> Since this is being rewritten, a few things can be fixed in this PR too :P

Which things do you mean?

File: src/DynoxHost.ts [hunk 3]
```diff
@@ -1,110 +1,110 @@
...
+        if (res.status !== 200) throw new Error(`[DH.js] An unexpected error has occurred`);
+        let json = res.json();
+        return json;
+    }
+
+    async setPowerState(id: string, state: string) {
+        /**
+         * Set server power state.
+         * @param {String} id Server ID
+         * @param {String} state Power State
```
Comment (devnote-dev) [inline src/DynoxHost.ts]: and here

File: src/DynoxHost.ts [hunk 4]
```diff
@@ -1,110 +1,110 @@
...
+        });
+        if (res.status === 404) throw new Error(`[DH.js] Server (${id}) not found`);
+        if (res.status !== 204) throw new Error(`[DH.js] An unexpected error has occurred`);
+        return { output: 'success' };
+    }
+
+    async createBackup(id: string) {
+        /**
+         * Creates a backup.
+         * @param {String} id Server ID
```
Comment (devnote-dev) [inline src/DynoxHost.ts]: here too

File: src/DynoxHost.ts [hunk 5]
```diff
@@ -1,110 +1,110 @@
...
+        if (res.status !== 200) throw new Error('[DH.js] An unexpected error has occurred');
+        let json = res.json();
+        return json;
+    }
+
+    async getBackupDetails(id: string, backupID: string) {
+        /**
+         * Gets details about a backup.
+         * @param {String} id Server ID
+         * @param {String} backupID Backup ID
```
Comment (devnote-dev) [inline src/DynoxHost.ts]: and also here

File: src/DynoxHost.ts [hunk 6]
```diff
@@ -1,110 +1,110 @@
...
+        if (!backupID) throw new Error('[DH.js] Missing Backup ID in getBackupDetails()');
+        if (typeof backupID !== 'string') throw new Error('[DH.js] Backup ID in getBackupDetails() must be a String');
+
+        let res = await fetch(`https://panel.dynox.us/api/client/servers/${id}/backups/${backupID}`, {
+            method: 'GET',
+            headers: { 'Authorization': `Bearer ${this.apikey}`, 'Content-Type': 'application/json', 'Accept': 'application/json' }
+        });
+        if (res.status === 404) throw new Error(`[DH.js] Server (${id}) or Backup ${backupID} not found`);
+        if (res.status !== 200) throw new Error('[DH.js] An unexpected error has occurred');
+        let json = res.json();
```
Comment (devnote-dev) [inline src/DynoxHost.ts:83]: this can be shortened to `return res.json();`

**Review by devnote-dev — CHANGES_REQUESTED** [review]

**Comment by tovade** [comment]
I can remove them as typescript auto will show them

**Comment by tovade** [comment]
ready to merge 👍

**Review by devnote-dev — APPROVED** [review]

**Pull request closed without merging**